src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,612 @@
+                int24 minTick = TickMath.minUsableTick(key.tickSpacing);
+                int24 maxTick = TickMath.maxUsableTick(key.tickSpacing);
+                if (params.tickLower == minTick || params.tickUpper == maxTick) {
+                    storedPosition.tickLower = params.tickLower;
+                    storedPosition.tickUpper = params.tickUpper;
+                    storedPosition.salt = params.salt;
+                    if (liquidityDelta > 0) {
+                        storedPosition.liquidity = uint128(uint256(liquidityDelta));
+                    }
+                }
+            } else if (storedPosition.salt == params.salt) {
+                if (liquidityDelta > 0) {
+                    storedPosition.liquidity += uint128(uint256(liquidityDelta));
+                } else {
+                    storedPosition.liquidity -= uint128(uint256(-liquidityDelta));
+                }
+            }
+        }
+
+        emit ModifyLiquidity(key, params);
+        return (BaseHook.afterAddLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+    }
+
+    /// @inheritdoc BaseHook
+    function _afterRemoveLiquidity(
+        address,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) internal override returns (bytes4, BalanceDelta) {
+        emit ModifyLiquidity(key, params);
+        return (BaseHook.afterRemoveLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+    }
+
+    /// @inheritdoc BaseHook
+    function _beforeSwap(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.SwapParams calldata params,
+        bytes calldata
+    ) internal override returns (bytes4, BeforeSwapDelta, uint24) {
+        if (sender == address(this)) {
+            return (BaseHook.beforeSwap.selector, BeforeSwapDeltaLibrary.ZERO_DELTA, 0);
+        }
+
+        uint24 fee = getFees[key.toId()].customFee;
+
+        uint256 swapAmount =
+            params.amountSpecified < 0 ? uint256(-params.amountSpecified) : uint256(params.amountSpecified);
+
+        if (fee == 0 || swapAmount == 0) {
+            return (BaseHook.beforeSwap.selector, BeforeSwapDeltaLibrary.ZERO_DELTA, 0);
+        }
+
+        uint256 feeAmount = FullMath.mulDiv(swapAmount, fee, MAX_SWAP_FEE);
+        if (feeAmount == 0) {
+            return (BaseHook.beforeSwap.selector, BeforeSwapDeltaLibrary.ZERO_DELTA, 0);
+        }
+
+        Currency feeCurrency = (params.amountSpecified < 0) == params.zeroForOne ? key.currency0 : key.currency1;
+
+        poolManager.take(feeCurrency, address(this), feeAmount);
+
+        if (feeCurrency == key.currency0) {
+            getFees[key.toId()].fees0 += uint128(feeAmount);
+        } else {
+            getFees[key.toId()].fees1 += uint128(feeAmount);
+        }
+
+        BeforeSwapDelta returnDelta = toBeforeSwapDelta(int128(int256(feeAmount)), 0);
```

Comment:
Need to revisit this to confirm but I think we should pass a negative value in the exact output case:
- Exact input: We add `feeAmount` to negative `amountSpecified`, decreasing the input amount
  - User pays the same input since the fee is taken but gets a lesser computed output
- Exact output: We add `feeAmount` to positive `amountSpecified`, increasing the output amount
  - User needs a larger input to cover larger output but also has to cover the amount taken for the fee

Reply:
On second thought, I think this is correct. In the exact output case the user receives the expected output amount but pays for an additional fee amount which is taken